### PR TITLE
refactor(storage): enable tcp keepalive and timeout for s3 client

### LINKED ISSUE #10209: bug: compaction task stuck for hours
### Describe the bug

Rencently we have seen a small cg3 (MV compaction group) L0->L0 compaction task getting stuck for hours, which cause L0->base level compaction to block for the corresponding compaction group. Some findings:
- The stuck compaction task contains 60 SSTs with total size=~13MB so it is a relatively small task.
- Compactor is alive and can handle tasks from both the same and different compaction groups as usual.
- We saw "Ready to handle compaction task" [log](https://github.com/risingwavelabs/risingwave/blob/572780bbca08c25a5d2f71016b915733e146df04/src/storage/src/hummock/compactor/mod.rs#L279) but not "Finished compaction task" [log](https://github.com/risingwavelabs/risingwave/blob/572780bbca08c25a5d2f71016b915733e146df04/src/storage/src/hummock/compactor/mod.rs#L374). This means compactor indeed received the task but not finished it.
- The task progress and heartbeat of the stuck task is continously reported to meta node. We use a [guard](https://github.com/risingwavelabs/risingwave/blob/572780bbca08c25a5d2f71016b915733e146df04/src/storage/src/hummock/compactor/mod.rs#L257) to make sure the task progress is cleared if the task finishes or errors out so that means the task is indeed stuck in the compactor side.
- Metric "Compacting SSTable Count" reported by meta node stays at 60 for cg3 level0. This means meta indeed didn't receive ReportCompactionTasksRequest from the compactor.

### To Reproduce

_No response_

### Expected behavior

_No response_

### Additional context

image
image
image

[log.csv](https://github.com/risingwavelabs/risingwave/files/11672843/log.csv)

### PR DESCRIPTION
**Pull request by zwang28**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR makes following changes to object store crate:

- Enable tcp keepalive for S3 client.
- Add `object_store_streaming_read_timeout_ms/object_store_streaming_upload_timeout_ms/object_store_upload_timeout_ms/object_store_read_timeout_ms` for object store client.
- Default timeout for those configs is set to 10 minutes or 60 minutes conservatively. All of them are configurable via config file.

Note that `object_store_read_timeout_ms` is also reused for lightweight op including list/delete/meta_data.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

## Documentation

- [ ] My PR contains user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/object_store/src/object/s3.rs [hunk 1]
```diff
@@ -559,6 +560,46 @@ impl ObjectStore for S3ObjectStore {
...
 }
 
+/// The connector is exactly the same as the default one except `set_keepalive`.
+pub fn customized_connector(
+    settings: &aws_smithy_client::http_connector::ConnectorSettings,
+    sleep: Option<Arc<dyn AsyncSleep>>,
+) -> Option<aws_smithy_client::erase::DynConnector> {
+    let default_hyper_builder = || -> hyper::client::Builder {
+        let builder = hyper::client::Builder::default();
+        // builder.pool_max_idle_per_host(0);
```
Comment (Li0k) [inline src/object_store/src/object/s3.rs]: remove this

**Comment by zwang28** [comment]
One thing worth noticing, after set SO_KEEPALIVE, I don't observe any blocked iter ever. However I had expected to see some blocked iter and subsequently they are unblocked by keep alive. This is still unexplained.

File: src/object_store/src/object/s3.rs [hunk 2]
```diff
@@ -591,8 +627,17 @@ impl S3ObjectStore {
         }
 
         // Retry 3 times if we get server-side errors or throttling errors
-        let sdk_config_loader =
+        let mut sdk_config_loader =
             aws_config::from_env().retry_config(RetryConfig::standard().with_max_attempts(4));
+        #[cfg(not(madsim))]
+        {
+            // Customize connector
```
Comment (Li0k) [inline src/object_store/src/object/s3.rs:634]:
```rust
    let mut http = hyper::client::HttpConnector::new();
        http.set_keepalive(Some(Duration::from_secs(600)));
        http.enforce_http(false);
        let smithy_connector = aws_smithy_client::hyper_ext::Adapter::builder()
            .hyper_builder(hyper::client::Builder::default())
            .connector_settings(ConnectorSettings::builder().build())
            .build(http);
        // Retry 3 times if we get server-side errors or throttling errors
        let sdk_config_loader = aws_config::from_env()
            .retry_config(RetryConfig::standard().with_max_attempts(4))
            .http_connector(smithy_connector);
```
We have not changed ConnectorSettings, can we use the above code to simplify customized_connector?

**Review by wcy-fdu — APPROVED** [review]
LGTM!

File: src/object_store/src/object/mod.rs [hunk 3]
```diff
@@ -442,9 +452,19 @@ pub struct MonitoredObjectStore<OS: ObjectStore> {
 ///   - `failure-count`
 impl<OS: ObjectStore> MonitoredObjectStore<OS> {
     pub fn new(store: OS, object_store_metrics: Arc<ObjectStoreMetrics>) -> Self {
+        let streaming_read_bytes_timeout = Duration::from_millis(from_env_or_default(
+            "RW_OBJECT_STORE_STREAMING_READ_BYTES_TIMEOUT_MS",
```
Comment (hzxa21) [inline src/object_store/src/object/mod.rs:456]: shall we make this part of the storage config?

File: src/object_store/src/object/mod.rs [hunk 4]
```diff
@@ -508,17 +528,20 @@ impl<OS: ObjectStore> MonitoredObjectStore<OS> {
...
-            .read(path, block_loc)
-            .verbose_instrument_await("object_store_read")
-            .await
-            .map_err(|err| {
-                ObjectError::internal(format!(
-                    "read {:?} in block {:?} failed, error: {:?}",
-                    path, block_loc, err
-                ))
-            });
+        let res = tokio::time::timeout(self.read_timeout, async {
```
Comment (hzxa21) [inline src/object_store/src/object/mod.rs:531]: Do we also need a timeout for upload and streaming upload in MonitoredObjectStore?

File: src/object_store/src/object/mod.rs [hunk 5]
```diff
@@ -399,10 +404,13 @@ impl MonitoredStreamingReader {
...
         self.operation_size += data_len;
-        let ret =
+        let ret = tokio::time::timeout(self.streaming_read_bytes_timeout, async {
             self.inner.read_exact(buf).await.map_err(|err| {
                 ObjectError::internal(format!("read_bytes failed, error: {:?}", err))
-            });
+            })
+        })
+        .await
+        .unwrap_or_else(|_| Err(ObjectError::internal("read_bytes timeout")));
```
Comment (Li0k) [inline src/object_store/src/object/mod.rs:413]: Can we expose the `streaming_read_bytes_timeout` directly in Error ?,so it's easier to find the problem?

File: src/object_store/src/object/mod.rs (hunk 4 above)
Comment (Li0k) [reply src/object_store/src/object/mod.rs:531]: +1 for upload and streaming upload

File: src/object_store/src/object/mod.rs [hunk 6]
```diff
@@ -508,17 +528,20 @@ impl<OS: ObjectStore> MonitoredObjectStore<OS> {
...
+                .await
+                .map_err(|err| {
+                    ObjectError::internal(format!(
+                        "read {:?} in block {:?} failed, error: {:?}",
+                        path, block_loc, err
+                    ))
+                })
+        })
+        .await
+        .unwrap_or_else(|_| Err(ObjectError::internal("read timeout")));
```
Comment (Li0k) [inline src/object_store/src/object/mod.rs:544]: ditto

File: src/object_store/src/object/mod.rs [hunk 7]
```diff
@@ -545,11 +568,14 @@ impl<OS: ObjectStore> MonitoredObjectStore<OS> {
...
-            .verbose_instrument_await("object_store_readv")
-            .await;
+        let res = tokio::time::timeout(self.read_timeout, async {
+            self.inner
+                .readv(path, block_locs)
+                .verbose_instrument_await("object_store_readv")
+                .await
+        })
+        .await
+        .unwrap_or_else(|_| Err(ObjectError::internal("readv timeout")));
```
Comment (Li0k) [inline src/object_store/src/object/mod.rs:578]: ditto

**Review by Li0k — APPROVED** [review]
Rest LGTM, thanks for the effort

File: src/object_store/src/object/mod.rs (hunk 4 above)
Comment (wcy-fdu) [reply src/object_store/src/object/mod.rs:531]: Out of topic, do we need to set speed for write(upload)? Speed is also considered as part of timeout in OpenDAL, for example, if there is no 1024 bytes per second, aka, 1KiB/s, the write will also return TimeoutError.
>For Writer::write(vec![1024*1024*1024])
didn’t return in 60s, it’s ok, we will keep waiting.
didn’t return in 1024s (1GiB/1MiB), we will return a timeout error.

File: src/object_store/src/object/mod.rs (hunk 5 above)
Comment (zwang28) [reply src/object_store/src/object/mod.rs:413]: ObjectErrorInner is classified by underlying impl + an Internal. I think it's weird to add another error kind.
```
enum ObjectErrorInner {
    #[error(transparent)]
    S3(BoxedError),
    #[error("disk error: {msg}")]
    Disk {
        msg: String,
        #[source]
        inner: io::Error,
    },
    #[error(transparent)]
    Opendal(opendal::Error),
    #[error(transparent)]
    Mem(crate::object::mem::Error),
    #[error("Internal error: {0}")]
    Internal(String),
}
```

**Review by Li0k — APPROVED** [review]
LGTM

**Review by hzxa21 — APPROVED** [review]

**Review by wcy-fdu — APPROVED** [review]

**Pull request merged**